File: koi/scanner/scanner.go
package scanner

import (
	"fmt"
	"strings"

	"github.com/jesperkha/koi/koi/token"
	"github.com/jesperkha/koi/koi/util"
)

type Scanner struct {
	// Number of syntax errors encountered and reported to handler.
	NumErrors int

	errors util.ErrorHandler
	file   *token.File

	src       []byte // Source text to scan
	row       int    // Current line
	col       int    // Current column, resets for each newline
	startCol  int    // Column of first character in token
	pos       int    // Pointer to currently scanned character
	base      int    // Pointer to first character of currently scanned token
	lineBegin int    // First character on current line
}

// New makes a new Scanner object for the given file. The src is the raw text
// input to scan. Scanner only accepts ascii text.
func New(file *token.File) *Scanner {
	return &Scanner{
		file:   file,
		src:    file.Src,
		errors: util.ErrorHandler{},
	}
}

// Scan consumes the next token and returns it, advancing the Scanner.
func (s *Scanner) Scan() token.Token {
	if s.eof() {
		return s.eofToken()
	}

	tok := s.scanWhitespace()

	s.base = s.pos
	s.startCol = s.col
	return tok
}

// ScanAll scans input source until everything is tokenized.
func (s *Scanner) ScanAll() []token.Token {
	toks := []token.Token{}
	var t token.Token

	for !t.Eof {
		t = s.Scan()
		toks = append(toks, t)
	}

	return toks
}

func (s *Scanner) Error() error {
	return s.errors.Error()
}

func isAlpha(c byte) bool {
	return strings.Contains("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ_", string(c))
}

func isNum(c byte) bool {
	return strings.Contains("0123456789", string(c))
}

func isWhitespace(c byte) bool {
	return strings.Contains("\t\r ", string(c))
}

func (s *Scanner) eofToken() token.Token {
	return token.Token{
		Type:   token.EOF,
		Eof:    true,
		Pos:    s.tokenPos(),
		EndPos: s.tokenEndPos(),
	}
}

func (s *Scanner) err(f string, args ...any) {
	msg := fmt.Sprintf(f, args...)
	s.errors.Pretty(s.row+1, s.file.Line(s.row), msg, s.startCol, s.col)
	s.NumErrors++
}

// peek next byte in input. Return 0 on EOF.
func (s *Scanner) peek() byte {
	if s.pos+1 >= len(s.src) {
		return 0
	}
	return s.src[s.pos+1]
}

func (s *Scanner) next() {
	if s.eof() {
		return
	}

	s.col++
	if s.cur() == '\n' {
		s.row++
		s.col = 0
		s.startCol = 0
		s.base = s.pos + 1
		s.lineBegin = s.pos + 1
	}

	s.pos++
}

func (s *Scanner) cur() byte {
	if s.eof() {
		return 0
	}
	return s.src[s.pos]
}

func (s *Scanner) eof() bool {
	return s.pos >= len(s.src)
}

func (s *Scanner) interval() string {
	return string(s.src[s.base:s.pos])
}

// Get the current base position of the Scanner as a token position.
func (s *Scanner) tokenPos() token.Pos {
	return token.Pos{
		Col:       s.startCol,
		Row:       s.row,
		Offset:    s.base,
		File:      s.file,
		LineBegin: s.lineBegin,
	}
}

// Get the current position of the Scanner as a token position.
func (s *Scanner) tokenEndPos() token.Pos {
	return token.Pos{
		Col:       s.col,
		Row:       s.row,
		Offset:    s.pos,
		File:      s.file,
		LineBegin: s.lineBegin,
	}
}

func (s *Scanner) scanWhitespace() token.Token {
	for !s.eof() && isWhitespace(s.cur()) {
		s.next()
		s.base = s.pos
		s.startCol = s.col
	}

	if s.eof() {
		return s.eofToken()
	}

	return s.scanNewline()
}

func (s *Scanner) scanNewline() token.Token {
	if s.cur() == '\n' {
		pos := s.tokenPos()
		endPos := s.tokenEndPos()
		endPos.Col++
		endPos.Offset++
		s.next()
		return token.Token{
			Type:   token.NEWLINE,
			Lexeme: "NEWLINE",
			Pos:    pos,
			EndPos: endPos,
		}
	}

	return s.scanComment()
}

func (s *Scanner) scanComment() token.Token {
	if s.cur() != '/' || s.peek() != '/' {
		return s.scanIdentifier()
	}

	for !s.eof() && s.cur() != '\n' {
		s.next()
	}

	// Newline tokens after comments have no
	// semantic value so we can just ignore them.
	s.next()
	return s.scanWhitespace()
}

func (s *Scanner) scanIdentifier() token.Token {
	if !isAlpha(s.cur()) {
		return s.scanNumber()
	}

	for !s.eof() && (isAlpha(s.cur()) || isNum(s.cur())) {
		s.next()
	}

	str := s.interval()
	typ := token.IDENT

	if t, ok := token.Keywords[str]; ok {
		typ = t
	}

	return token.Token{
		Type:   typ,
		Lexeme: str,
		Pos:    s.tokenPos(),
		EndPos: s.tokenEndPos(),
	}
}

func (s *Scanner) scanNumber() token.Token {
	if !isNum(s.cur()) {
		return s.scanString()
	}

	dots := 0

	for !s.eof() {
		if s.cur() == '.' {
			dots++
		} else if !isNum(s.cur()) {
			break
		}

		s.next()
	}

	if dots > 1 {
		s.err("number literal can have at most one decimal point")
	}

	typ := token.INT_LIT
	if dots > 0 {
		typ = token.FLOAT_LIT
	}

	return token.Token{
		Type:    typ,
		Lexeme:  s.interval(),
		Invalid: dots > 1,
		Float:   dots > 0,
		Pos:     s.tokenPos(),
		EndPos:  s.tokenEndPos(),
	}
}

func (s *Scanner) scanString() token.Token {
	if s.cur() != '"' {
		return s.scanByteString()
	}

	s.next()        // Consume start quote
	closed := false // True if found end quote on current line

	for !s.eof() {
		if s.cur() == '"' {
			s.next() // Consume end quote
			closed = true
			break
		}

		if s.cur() == '\n' {
			break
		}

		s.next()
	}

	if !closed {
		s.err("string literals must have a terminating quote on the same line")
	}

	return token.Token{
		Type:    token.STRING_LIT,
		Lexeme:  s.interval(),
		Invalid: !closed,
		Pos:     s.tokenPos(),
		EndPos:  s.tokenEndPos(),
	}
}

func (s *Scanner) scanByteString() token.Token {
	if s.cur() != '\'' {
		return s.scanSymbol()
	}

	s.next()        // Consume start quote
	closed := false // True if found end quote on current line

	length := 0
	for !s.eof() {
		if s.cur() == '\'' {
			s.next() // Consume end quote
			closed = true
			break
		}

		if s.cur() == '\n' {
			break
		}

		s.next()
		length++
	}

	if !closed {
		s.err("byte literals must have a terminating quote on the same line")
	}

	if length != 1 {
		s.err("byte literals must have exactly one character")
	}

	return token.Token{
		Type:    token.BYTE_LIT,
		Lexeme:  s.interval(),
		Invalid: !closed || length != 1,
		Pos:     s.tokenPos(),
		EndPos:  s.tokenEndPos(),
	}
}

func (s *Scanner) scanSymbol() token.Token {
	for sym, typ := range token.DoubleSymbols {
		if s.cur() != sym[0] {
			continue
		}

		if s.peek() == sym[1] {
			s.next()
			s.next()

			return token.Token{
				Type:   typ,
				Lexeme: s.interval(),
				Pos:    s.tokenPos(),
				EndPos: s.tokenEndPos(),
			}
		}
	}

	if typ, ok := token.SingleSymbols[string(s.cur())]; ok {
		s.next()
		return token.Token{
			Type:   typ,
			Lexeme: s.interval(),
			Pos:    s.tokenPos(),
			EndPos: s.tokenEndPos(),
		}
	}

	return s.scanIllegal()
}

func (s *Scanner) scanIllegal() token.Token {
	s.next()
	s.err("illegal token")

	return token.Token{
		Type:    token.ILLEGAL,
		Invalid: true,
		Lexeme:  s.interval(),
		Pos:     s.tokenPos(),
		EndPos:  s.tokenEndPos(),
	}
}

File: src/token/scanner.rs
use tracing::{debug, info, trace};

use crate::{
    config::Config,
    error::{Diagnostics, Report, Res},
    token::{Pos, Source, Token, TokenKind, str_to_token},
};

pub fn scan(src: &Source, config: &Config) -> Res<Vec<Token>> {
    let scanner = Scanner::new(src, config);
    scanner.scan()
}

struct Scanner<'a> {
    source: &'a Source,
    pos: usize,
    row: usize,
    col: usize,
    line_begin: usize,
    _config: &'a Config,
    diag: Diagnostics,
}

impl<'a> Scanner<'a> {
    fn new(source: &'a Source, config: &'a Config) -> Self {
        Scanner {
            _config: config,
            source,
            pos: 0,
            col: 0,
            row: 0,
            line_begin: 0,
            diag: Diagnostics::new(),
        }
    }

    fn scan(mut self) -> Res<Vec<Token>> {
        info!("Scanning file: {}", self.source.filepath);

        // No input
        if self.eof() {
            info!("No input");
            return Ok(Vec::new());
        }

        while !self.eof() {
            match self.scan_all() {
                // If ok and we did not encounter errors before, return result
                // Otherwise ignore result as one or more errors have been raised
                Ok(toks) => {
                    if self.diag.is_empty() {
                        debug!("success: {} tokens", toks.len());
                        return Ok(toks);
                    }
                }
                // If error add to set and skip to next 'safe' spot
                Err(err) => {
                    self.diag.add(err);
                    self.pos += self.peek_while(|p| !Scanner::is_whitespace(p))
                }
            }
        }

        info!("Fail: finished with {} errors", self.diag.num_errors());
        Err(self.diag)
    }

    fn scan_all(&mut self) -> Result<Vec<Token>, Report> {
        let mut tokens = Vec::new();

        while !self.eof() {
            let (token, consumed) = match self.cur() {
                // Whitespace tokens are ignored and not added to token list
                v if Scanner::is_whitespace(v) => (
                    Token::new(TokenKind::Whitespace, 0, self.pos()),
                    self.peek_while(Scanner::is_whitespace),
                ),

                // Line comment
                b'/' if matches!(self.peek(), Some(b'/')) => {
                    let len = self.peek_while(|b| b != b'\n');
                    let lexeme = self.source.str_range(self.pos, self.pos + len);
                    let tok = Token::new(TokenKind::Comment(lexeme.to_owned()), len, self.pos());
                    (tok, len)
                }

                // Block comment
                b'/' if matches!(self.peek(), Some(b'*')) => {
                    let mut depth = 1;
                    let mut i = self.pos + 2; // Skip opening /*

                    while i + 1 < self.len() && depth > 0 {
                        let c1 = self.at(i);
                        let c2 = self.at(i + 1);

                        if c1 == b'/' && c2 == b'*' {
                            depth += 1;
                            i += 2;
                            continue;
                        } else if c1 == b'*' && c2 == b'/' {
                            depth -= 1;
                            i += 2;
                            continue;
                        }

                        i += 1;
                    }

                    if depth != 0 {
                        return Err(Report::new_length(
                            "block comment was not terminated",
                            &self.pos(),
                            2,
                        ));
                    }

                    (
                        Token::new(TokenKind::Whitespace, 0, self.pos()),
                        i - self.pos,
                    )
                }

                // Newline character resets the row and col.
                b'\n' => {
                    let pos = self.pos();
                    self.row += 1;
                    self.col = 0;
                    self.line_begin = self.pos + 1;
                    (Token::new(TokenKind::Newline, 1, pos), 1)
                }

                // Identifier or keyword
                v if Scanner::is_alpha(v) => {
                    let length = self.peek_while(Scanner::is_alphanum);
                    let lexeme = self.source.str_range(self.pos, self.pos + length);

                    if let Some(k) = str_to_token(lexeme) {
                        (Token::new(k.clone(), length, self.pos()), length)
                    } else {
                        (
                            Token::new(TokenKind::IdentLit(lexeme.to_owned()), length, self.pos()),
                            length,
                        )
                    }
                }

                // Number
                v if Scanner::is_number(v) => {
                    let mut length = self.peek_while(Scanner::is_numeric);
                    let mut lexeme = self.source.str_range(self.pos, self.pos + length);

                    // Ignore ending period for now, checked by Checker
                    if lexeme.ends_with(".") {
                        length -= 1;
                        lexeme = lexeme.trim_end_matches(".");
                    }

                    let kind = if lexeme.contains('.') {
                        match lexeme.parse() {
                            Ok(f) => TokenKind::FloatLit(f),
                            _ => return Err(self.error("invalid number literal", length)),
                        }
                    } else {
                        match lexeme.parse() {
                            Ok(f) => TokenKind::IntLit(f),
                            _ => return Err(self.error("invalid number literal", length)),
                        }
                    };

                    (Token::new(kind, length, self.pos()), length)
                }

                // String
                b'"' => self.scan_string(b'"')?,

                // Byte string
                b'\'' => {
                    let (tokens, length) = self.scan_string(b'\'')?;
                    if length != 3 {
                        return Err(self.error("byte string must be exactly one character", length));
                    }

                    (tokens, length)
                }

                // Match either one or two tokens (single/double symbol)
                _ => {
                    let try_match = |len| {
                        let lexeme = self.source.str_range(self.pos, self.pos + len);
                        str_to_token(&lexeme)
                            .map(|kind| Token::new(kind.to_owned(), len, self.pos()))
                    };

                    if let Some(token) = self
                        .peek()
                        .filter(|&c| !Scanner::is_alphanum(c))
                        .and_then(|_| try_match(2))
                    {
                        (token, 2)
                    } else if let Some(token) = try_match(1) {
                        (token, 1)
                    } else {
                        return Err(self.error("illegal token", 1));
                    }
                }
            };

            trace!("consumed token: '{}'", token);
            self.pos += consumed;

            // Col must not advance after a newline. It is reset to 0 above and must remain 0
            // before next iteration. Incrementing now would cause the first token on the new
            // line to have col=1
            if !token.kind.eq(&TokenKind::Newline) {
                self.col += consumed;
            }

            if !token.kind.eq(&TokenKind::Whitespace) {
                tokens.push(token);
            }
        }

        Ok(tokens)
    }

    fn pos(&self) -> Pos {
        Pos {
            source_id: self.source.id,
            row: self.row,
            col: self.col,
            offset: self.pos,
            line_begin: self.line_begin,
        }
    }

    fn at(&self, pos: usize) -> u8 {
        assert!(
            pos < self.len(),
            "tried to access pos {} when src is {}",
            pos,
            self.len()
        );
        self.source.src[pos]
    }

    fn peek(&self) -> Option<u8> {
        if self.pos + 1 >= self.len() {
            None
        } else {
            Some(self.at(self.pos + 1))
        }
    }

    fn cur(&self) -> u8 {
        self.source.src[self.pos]
    }

    fn eof(&self) -> bool {
        self.pos >= self.len()
    }

    fn len(&self) -> usize {
        self.source.src.len()
    }

    fn error(&self, msg: &str, length: usize) -> Report {
        Report::new_length(msg, &self.pos(), length)
    }

    /// Peeks tokens while predicate returns true. Returns number of tokens peeked.
    fn peek_while<P>(&mut self, predicate: P) -> usize
    where
        P: Fn(u8) -> bool,
    {
        let mut consumed = 0;
        while self.pos + consumed < self.len() && predicate(self.source.src[self.pos + consumed]) {
            consumed += 1;
        }

        consumed
    }

    /// Scans a string literal, starting at the current position.
    fn scan_string(&mut self, quote: u8) -> Result<(Token, usize), Report> {
        self.pos += 1;
        let mut length = self.peek_while(|b| b != quote && b != b'\n');
        self.pos -= 1;

        // Was string actually closed?
        let check_pos = self.pos + length + 1; // Of end quote
        if check_pos >= self.len() || self.at(check_pos) != quote {
            let mut pos = self.pos();
            pos.col += check_pos;
            pos.offset += check_pos;
            return Err(Report::new_length("expected end quote", &pos, 1));
        }

        length += 2; // Include start and end quote
        let lexeme = self.source.str_range(self.pos + 1, self.pos + length - 1);

        Ok((
            Token::new(TokenKind::StringLit(lexeme.to_string()), length, self.pos()),
            length,
        ))
    }

    fn is_number(n: u8) -> bool {
        n >= b'0' && n <= b'9'
    }

    fn is_numeric(n: u8) -> bool {
        Scanner::is_number(n) || n == b'.'
    }

    fn is_whitespace(b: u8) -> bool {
        b == b' ' || b == b'\t' || b == b'\r'
    }

    fn is_alpha(b: u8) -> bool {
        (b >= b'a' && b <= b'z') || (b >= b'A' && b <= b'Z') || b == b'_'
    }

    fn is_alphanum(b: u8) -> bool {
        Scanner::is_alpha(b) || Scanner::is_number(b)
    }
}
